src/internal/easee/auth.go

```diff
@@ -61,15 +96,73 @@ func (a *authenticator) AccessToken() (string, error) {
 		return credentials.AccessToken, nil
 	}
 
+	if a.status == statusWaitingToReconnect || a.status == statusConnectionFailed {
+		return "", errors.New("connection interrupted, waiting for user to re-login")
+	}
+
 	newCredentials, err := a.http.RefreshToken(credentials.AccessToken, credentials.RefreshToken)
 	if err != nil {
-		return "", fmt.Errorf("failed to fetch new credentials: %w", err)
+		return "", a.handleFailedRefreshToken(err)
 	}
 
+	// reset backoff stats
+	a.status = statusWorkingProperly
+	a.attempts = 0
+
 	err = a.cfgSvc.SetCredentials(newCredentials.AccessToken, newCredentials.RefreshToken, newCredentials.ExpiresIn)
 	if err != nil {
 		return "", fmt.Errorf("failed to save credentials in storage: %w", err)
 	}
 
 	return newCredentials.AccessToken, nil
 }
+
+func (a *authenticator) Logout() error {
+	return a.cfgSvc.ClearCredentials()
+}
+
+// handleFailedRefreshToken sets a correct status when refresh operation has failed.
+// relies on mutex protection within a caller.
+func (a *authenticator) handleFailedRefreshToken(err error) error {
+	// we aren't able to refresh anymore. Requires user re-login
+	var httpError HTTPError
+	if ok := errors.As(err, &httpError); ok && httpError.Status == 401 {
+		_ = a.notificationManager.Event(&notification.Event{EventName: notificationEaseeStatusOffline})
+		a.status = statusConnectionFailed
+		_ = a.cfgSvc.ClearCredentials()
+
+		return fmt.Errorf("received unauthorized error, re-login is required. %w", err)
+	}
+
+	switch a.status { //nolint
+	case statusReconnecting:
+		if a.attempts == a.maxAttempts {
+			_ = a.notificationManager.Event(&notification.Event{EventName: notificationEaseeStatusOffline})
+			a.status = statusConnectionFailed
+
+			return fmt.Errorf("failed delayed attempt to refresh token. Re-login required. %w", err)
+		}
+
+		go a.hookResetToReconnecting()
+		a.status = statusWaitingToReconnect
+
+		return fmt.Errorf("failed delayed attempt to refresh token. Trying again. %w", err)
+	case statusWorkingProperly:
+		a.status = statusWaitingToReconnect
+		go a.hookResetToReconnecting()
+
+		return fmt.Errorf("failed to refresh token. Suspending for 5 minutes. %w", err)
+	default:
+		return fmt.Errorf("invalid auth status when refreshing token: %d. Error: %w", a.status, err)
+	}
+}
+
+// hookResetToReconnecting resets status to statusReconnecting after a delay.
+// mist be called in a separate goroutine.
```

Comment:
There is a typo here, but why do we have to reset the attempts concurrently? Isn't there a chance of a burst from the enclosing goroutine while the count gets updated in this function? Maybe I got it wrong.

Reply:
We reset to the `reconnectingState` concurrently because we don't want to block the current flow from returning an error on failed `refreshToken` request. 
We update `attempts` because it's rather comfortable to switch it at one place, together with the state.
Until the state isn't [changed to reconnecting or working-properly](https://github.com/futurehomeno/edge-easee-adapter/pull/19/files#diff-27be6ad3ad9afdd44dc0961d377b5ea2e07199b082fb1883fd0031e3649c3231R99) we don't try to refresh token again and attempts count doesn't matter. It matters only when we retry again few line above in the switch block.